tests/unit_tests/output_selection.cpp

```diff
@@ -109,28 +109,48 @@ TEST(select_outputs, order)
     offset = n_outs += (n); \
   }
 
-TEST(select_outputs, gamma)
+void pick_outputs(size_t npicks, std::vector<uint64_t> offsets, size_t n_outs, tools::gamma_picker picker, double &median)
 {
-  std::vector<uint64_t> offsets;
-
-  MKOFFSETS(300000, 1);
-  tools::gamma_picker picker(offsets);
-  std::vector<double> ages(100000);
+  std::vector<double> ages(npicks);
   double age_scale = 120. * (offsets.size() / (double)n_outs);
   for (size_t i = 0; i < ages.size(); )
   {
     uint64_t o = picker.pick();
-    if (o >= n_outs)
+    bool acceptable_bad_pick = o == std::numeric_limits<uint64_t>::max();
+    if (acceptable_bad_pick)
       continue;
     ages[i] = (n_outs - 1 - o) * age_scale;
-    ASSERT_GE(ages[i], 0);
+    ASSERT_GE(ages[i], CRYPTONOTE_DEFAULT_TX_SPENDABLE_AGE * 120);
     ASSERT_LE(ages[i], offsets.size() * 120);
     ++i;
   }
-  double median = epee::misc_utils::median(ages);
-  MDEBUG("median age: " << median / 86400. << " days");
-  ASSERT_GE(median, 1.3 * 86400);
-  ASSERT_LE(median, 1.4 * 86400);
+  median = epee::misc_utils::median(ages);
+}
+
+// TODO: deterministic tests would be nice
+TEST(select_outputs, gamma)
+{
+  std::vector<uint64_t> offsets;
+
+  MKOFFSETS(300000, 1);
+  tools::gamma_picker picker(offsets);
+
+  size_t NPICKS = 100000;
+  double median = 0;
+  pick_outputs(NPICKS, offsets, n_outs, picker, median);
+
+  // expected median is ~115,100s, or 1.33 * 86400. Calculated by running the algorithm
+  // under the same conditions as the paper (outputs <10 blocks old can be selected, and
+  // are selected from chain tip)
+  double MEDIAN_LOWER_BOUND = 1.28 * 86400;
+  double MEDIAN_UPPER_BOUND = 1.38 * 86400;
+
+  // should be rare, but if needed, 10x NPICKS and try again to get a more accurate median
```

Comment:
What if you iteratively added new picks and instead made an assertion regarding convergence to the expected median? So, you'd keep the old picks, add new ones, re-compute the combined median, and expect that the median gets closer to target. You can add a tolerance to the convergence expectation such that, for example, a single non-converging iteration tries again, but two or three or whatever would fail. This way, you can abort a lot quicker on what are probable failures, but also not fail on what are really OK implementations.